### DISCUSSION TITLE
Add `cancelable` operator

### DISCUSSION THREAD

**Discussion by igorkamyshev** [sergeysova/patronum#154] [Ideas]
Some effects should be cancelled because of different events. So, I propose to add operator `cancellable`, for creation cancellable effect.

Reference: 
+ https://github.com/doasync/fry-fx — it handles cancellable effects, but depends on AbortController and requires a lot of boring work.

API proposition:

```ts
const internalFx = createEffect(({ onCancel, ...otherParams }) => {
  // ...
  onCancel(() => {
    // ...
  })
})

const cancel = createEvent()

// Cancelled by event
const fx1 = cancellable({ effect: internalFx, trigger: cancel })

const turnOff = createEvent()
const $enabled = createStore(true)
  .on(turnOff, () => false)

// Cancelled by store value changed to `false`
const fx1 = cancellable({ effect: internalFx, grant: $enabled })
```

With this API, we can reimplement the most parts of `fry-fx` and `effector-reeffect`.

What do you think of this proposal? If you are OK with it, I try to implement naive version of it in PR.

**Comment by sergeysova** [comment]
Maybe it is better to name it in style of effector fabric-methods: `createCancellable`

**Reply by igorkamyshev** [reply]
But it isn't fabric. My proposal have a form of an operator.

**Reply by zerobias** [reply]
`attach` doesn’t have `create-` suffix, so this is not a requirement. ability to pass units to parameters is more important

**Comment by yumauri** [comment]
Raw try to implement base `cancellable` (supports cancel only by event):
https://share.effector.dev/hC05veNs
It could be extended, but I’m not sure about `fork` support with this approach...

**Reply by AlexandrHoroshih** [reply]
It doesn't work with fork 🤔
My guess it is due to usage of Promice.race with both effect and non-effect async functions as params - this way effector loses tracks of original effect - and `allSettled` just never ends
https://share.effector.dev/Eef9ugIC

I did a very similiar (but a little sloppy) implementation for this case a couple months ago, which can work in forked scope
It is more like a fabric for cancellable effects, but can work with effects too

UPD: apparently, using an effect as a handler in this implementation leads to the same problem - allSettled never ends 🤔
Usual function works correctly. 
Actually, it is just not properly used onCancel callback 🤔 - it should also reject that Promise, so whole thing works correctly, it is user-side bug. Working example (promise rejected too): https://share.effector.dev/xnqn2l85

**Reply by yumauri** [reply]
Damn `fork`, it never works 😅
Issue here is with cleanup function, which just cancels timeout and original effect just never ends. Whereas `needFxCounter` field is decremented on `finally` event, which is never launched.
Little fast workaround for that, I can think of — use actual handler instead of effect in `Promise.race` → https://share.effector.dev/KHZnL26w
But this will not work with attached effects, I guess...

**Reply by yumauri** [reply]
One more issue with my code — single instance of cleanup function for cancellable effects. And in result, cleanup actually works just for latest effect call :(

**Reply by zerobias** [reply]
[request cancellation for effector 22 and fork api](https://share.effector.dev/BcvWG4l3)

**Comment by AlexandrHoroshih** [comment]
Possible implementation here, feel free to join discussion
https://github.com/effector/patronum/pull/168
